README.md
#Overview
Couldn't find a nice clean class that I liked to interact with Tumblr v2 API. Some help from [ryanmcgrath twython class](https://github.com/ryanmcgrath/twython), thanks a lot. Reading over some of that code helped a lot!
Hope this documentation explains everything you need to get started. Any questions feel free to email me or inbox me.

#Install through pip...
```
pip install python-tumblpy
```

#Authorization URL
*Get an authorization url for your user*

```python
# Without a dynamic callback url
# This will use the callback url specified in your app

# Go to http://www.tumblr.com/oauth/apps and click your
# app to find out your dynamic callback url
t = Tumblpy(app_key = '*your app key*',
            app_secret = '*your app secret*')
```

If you wish to have a dynamic callback url, specify ```callback_url``` when you initiate the class.

```python
t = Tumblpy(app_key = '*your app key*',
            app_secret = '*your app secret*'
            callback_url = 'http://example.com/callback/')
```

```python
auth_props = t.get_authentication_tokens()
auth_url = auth_props['auth_url']

oauth_token = auth_props['oauth_token']
oauth_token_secret = auth_props['oauth_token_secret']

Connect with Tumblr via: % auth_url
```

Once you click "Allow" be sure that there is a URL set up to handle getting finalized tokens and possibly adding them to your database to use their information at a later date.

#Handling the callback
```python
t = Tumblpy(app_key = '*your app key*',
            app_secret = '*your app secret*',
            oauth_token=session['tumblr_session_keys']['oauth_token'],
            oauth_token_secret=session['tumblr_session_keys']['oauth_token_secret'])

# In Django, you'd do something like
# oauth_verifier = request.GET.get('oauth_verifier')

oauth_verifier = *Grab oauth verifier from URL*
authorized_tokens = t.get_access_token(oauth_verifier)

final_oauth_token = authorized_tokens['oauth_token']
final_oauth_token_secret = authorized_tokens['oauth_token_secret']

# Save those tokens to the database for a later use?
```

#Getting some user information
```python
# Get the final tokens from the database or wherever you have them stored

t = Tumblpy(app_key = '*your app key*',
            app_secret = '*your app secret*',
            oauth_token=final_tokens['oauth_token'],
            oauth_token_secret=final_tokens['oauth_token_secret'])

# Print out the user info, let's get the first blog url...
blog_url = t.post('user/info')
blog_url = blog_url['user']['blogs'][0]['url']

# Let's get their blog posts for that URL
posts = t.get('posts', blog_url=blog_url)

# Uploading a photo post
```python
'''
Assuming you already have authenticated and wanted to grab your blog URL first if not, use the following function
'''
blog_url = t.post('user/info')
blog_url = blog_url['user']['blogs'][0]['url']  # Will look something like http://omglegit.tumblr.com

files = open('/path/to/file/image.png', 'rb')
print t.post('post', blog_url=blog_url, params={'type':'photo', 'caption': 'Test Caption'}, files=files)
```

print posts
```

HISTORY.rst
.. :changelog:

History
-------

1.1.4 (2016-02-08)
++++++++++++++++++

- Remove old api url string formatting.
- Added ``posts`` method to Tumblpy, see README for example.

1.1.3 (2016-01-17)
++++++++++++++++++

- Fix missing import

1.1.2 (2016-12-22)
++++++++++++++++++

- Fix missing import

1.1.1 (2016-05-12)
++++++++++++++++++

- Fix issue where blogs using https:// were being parsed wrong


1.1.0 (2016-30-04)
++++++++++++++++++

- Add following and dashboard API methods


1.0.5 (2015-08-13)
++++++++++++++++++

- Add support for ``proxies`` keyword for requests


1.0.4 (2015-01-15)
++++++++++++++++++

- Fix request token decode issue in Python 3


1.0.3 (2014-10-17)
++++++++++++++++++

- Unpin ``requests`` and ``requests-oauthlib`` versions in ``setup.py``


1.0.2 (2013-05-31)
++++++++++++++++++

- Made the hotfix for posting photos a little more hotfixy... fixed posting just regular posts (as well as photos)

1.0.1 (2013-05-29)
++++++++++++++++++

- Hotfix image uploading (not sure why we have to pass ``params`` AND ``data`` to the POST, hotfix for the time being...)
- Allow for ints and floats (and longs in Python 2) to be passed as parameters to Tumblpy Tumblr API functions


1.0.0 (2013-05-23)
++++++++++++++++++

- Changed internal Tumblpy API structure, but Tumblpy functions should still work as they did before
- Updated README with more clear examples
- Added LICENSE
- ``_split_params_and_files`` has been moved to ``helpers.py``
- All ``Tumblpy`` exceptions are found in ``exceptions.py``
- Removed ``pool_maxsize`` from ``Tumblpy.__init__`` because it wasn't being used
- Removed ``timeout`` parameter from all request methods for the time being
- Removed ``TumblpyTimeout`` Exception
- Moved ``callback_url`` parameter from ``Tumblpy.__init__`` to ``get_authentication_tokens``
- All authentication and API calls over HTTPS
- Dropped Python 2.5 support
- Full, transparent Python 3.3 support


test.py
__author__ = "Russell Hay"


tumblpy/api.py
import requests
from requests_oauthlib import OAuth1

from . import __version__
from .compat import json, parse_qsl, urlencode, urlparse
from .exceptions import TumblpyAuthError, TumblpyError
from .helpers import _split_params_and_files


class Tumblpy(object):
    def __init__(self, app_key=None, app_secret=None, oauth_token=None,
                 oauth_token_secret=None, headers=None, proxies=None):

        # Define some API URLs real quick
        self.base_api_url = 'https://api.tumblr.com'
        self.api_version = 'v2'
        self.api_url = '%s/%s/' % (self.base_api_url, self.api_version)

        # Authentication URLs
        self.request_token_url = 'https://www.tumblr.com/oauth/request_token'
        self.access_token_url = 'https://www.tumblr.com/oauth/access_token'
        self.authorize_url = 'https://www.tumblr.com/oauth/authorize'
        self.authenticate_url = 'https://www.tumblr.com/oauth/authorize'

        self.default_params = {'api_key': app_key}

        req_headers = {'User-Agent': 'Tumblpy v' + __version__}
        if headers:
            req_headers.update(headers)

        self.app_key = app_key
        self.app_secret = app_secret
        self.oauth_token = oauth_token
        self.oauth_token_secret = oauth_token_secret

        auth = None
        if self.app_key and self.app_secret:
            if not self.oauth_token and not self.oauth_token_secret:
                auth = OAuth1(self.app_key, self.app_secret)
            else:
                auth = OAuth1(self.app_key, self.app_secret,
                              self.oauth_token, self.oauth_token_secret)

        self.client = requests.Session()
        self.client.proxies = proxies
        self.client.headers = req_headers
        self.client.auth = auth

    def get_authentication_tokens(self, callback_url=None):
        """Returns a dict including an authorization URL (auth_url) to direct a user to

            :param callback_url: (optional) Url the user is returned to after they authorize your app (web clients only)
        """

        request_args = {}
        if callback_url:
            request_args['oauth_callback'] = callback_url

        response = self.client.get(self.request_token_url, params=request_args)

        if response.status_code != 200:
            raise TumblpyAuthError('Seems something couldn\'t be verified with your OAuth junk. Error: %s, Message: %s' % (response.status_code, response.content))

        res = response.content
        if isinstance( response.content, bytes ):
            res = res.decode()

        request_tokens = dict(parse_qsl(res))

        if not request_tokens:
            raise TumblpyError('Unable to decode request tokens.')

        auth_url_params = {
            'oauth_token': request_tokens['oauth_token'],
        }
        if callback_url:
            auth_url_params['oauth_callback'] = callback_url

        request_tokens['auth_url'] = self.authenticate_url + '?' + urlencode(auth_url_params)

        return request_tokens

    def get_authorized_tokens(self, oauth_verifier):
        """Returns authorized tokens after they go through the auth_url phase.
        """
        response = self.client.get(self.access_token_url,
                                   params={'oauth_verifier': oauth_verifier})

        res = response.content
        if isinstance( response.content, bytes ):
            res = res.decode()

        authorized_tokens = dict(parse_qsl(res))
        if not authorized_tokens:
            raise TumblpyError('Unable to decode authorized tokens.')

        return authorized_tokens

    def request(self, endpoint, method='GET', blog_url=None,
                extra_endpoints=None, params=None):
        params = params or {}
        method = method.lower()

        if not method in ('get', 'post'):
            raise TumblpyError('Method must be of GET or POST')

        url = self.api_url  # http://api.tumblr.com/v2/

        if blog_url is not None:
            # http://api.tumblr.com/v2/blog/blogname.tumblr.com/
            blog_url = urlparse(blog_url)

            url = '%sblog/%s/' % (
                self.api_url,
                blog_url.hostname if blog_url.hostname is not None else blog_url.path
            )

        url = '%s%s' % (url, endpoint)
        if extra_endpoints is not None:
            # In cases like:
            # http://api.tumblr.com/v2/blog/blogname.tumblr.com/posts/type/
            # 'type' is extra in the url & thought this was the best way
            # Docs: http://www.tumblr.com/docs/en/api/v2#posts

            url = '%s/%s' % (url, '/'.join(extra_endpoints))

        params, files = _split_params_and_files(params)
        params.update(self.default_params)

        func = getattr(self.client, method)
        try:
            if method == 'get':
                response = func(url, params=params, allow_redirects=False)
            else:
                kwargs = {'data': params, 'files': files, 'allow_redirects': False}
                if files:
                    kwargs['params'] = params
                response = func(url, **kwargs)
        except requests.exceptions.RequestException:
            raise TumblpyError('An unknown error occurred.')

        if response.status_code == 401:
            raise TumblpyAuthError('Error: %s, Message: %s' % (response.status_code, response.content))

        content = response.content.decode('utf-8')
        try:
            if endpoint == 'avatar':
                content = {
                    'response': {
                        'url': response.headers.get('location')
                    }
                }
            else:
                content = json.loads(content)
        except ValueError:
            raise TumblpyError('Unable to parse response, invalid JSON.')

        try:
            content = content.get('response', {})
        except AttributeError:
            raise TumblpyError('Unable to parse response, invalid content returned: %s' % content)

        if response.status_code < 200 or response.status_code > 301:
            error_message = ''
            if content and (content.get('errors') or content.get('error')):
                if 'errors' in content:
                    for error in content['errors']:
                        error_message = '%s ' % error
                elif 'error' in content:
                    error_message = content['error']

            error_message = (error_message or
                             'There was an error making your request.')
            raise TumblpyError(error_message, error_code=response.status_code)

        return content

    def get(self, endpoint, blog_url=None, extra_endpoints=None, params=None):
        return self.request(endpoint, blog_url=blog_url,
                            extra_endpoints=extra_endpoints, params=params)

    def post(self, endpoint, blog_url=None, extra_endpoints=None, params=None):
        return self.request(endpoint, method='POST', blog_url=blog_url,
                            extra_endpoints=extra_endpoints, params=params)

    def get_avatar_url(self, blog_url, size=64):
        size = [str(size)] or ['64']
        return self.get('avatar', blog_url=blog_url, extra_endpoints=size)

    def following(self, kwargs=None):
        """
        Gets the blogs that the current user is following.
        :param limit: an int, the number of likes you want returned
        :param offset: an int, the blog you want to start at, for pagination.

            # Start at the 20th blog and get 20 more blogs.
            client.following({'offset': 20, 'limit': 20})

        :returns: A dict created from the JSON response
        """
        return self.get('user/following', params=kwargs)

    def dashboard(self, kwargs=None):
        """
        Gets the dashboard of the current user
        example: dashboard = client.dashboard({'limit': '3'})


        :param limit: an int, the number of posts you want returned
        :param offset: an int, the posts you want to start at, for pagination.
        :param type:   the type of post you want to return
        :param since_id:  return only posts that have appeared after this ID
        :param reblog_info: return reblog information about posts
        :param notes_info:  return notes information about the posts

        :returns: A dict created from the JSON response
        """
        return self.get('user/dashboard', params=kwargs)

    def posts(self, blog_url, post_type=None, kwargs=None):
        """
        Gets a list of posts from a particular blog
        :param blog_url: a string, the blogname you want to look up posts
                         for. eg: codingjester.tumblr.com
        :param post_type:  the type of posts you want returned, e.g. video. If omitted returns all post types.
        :param limit: an int, the number of likes you want returned
        :param offset: an int, the blog you want to start at, for pagination.
        :returns: A dict created from the JSON response
        """
        url = 'posts'
        if post_type:
            url = '%s/%s' % (url, post_type)

        return self.get(url, blog_url=blog_url, params=kwargs)

    def __repr__(self):
        return u'<TumblrAPI: %s>' % self.app_key


examples/example.py
import sys

from tumblpy import Tumblpy

key = raw_input('App Consumer Key: ')
secret = raw_input('App Consumer Secret: ')

if not 'skip-auth' in sys.argv:
    t = Tumblpy(key, secret)

    callback_url = raw_input('Callback URL: ')

    auth_props = t.get_authentication_tokens(callback_url=callback_url)
    auth_url = auth_props['auth_url']

    OAUTH_TOKEN_SECRET = auth_props['oauth_token_secret']

    print('Connect with Tumblr via: {}'.format(auth_url))

    oauth_token = raw_input('OAuth Token (from callback url): ')
    oauth_verifier = raw_input('OAuth Verifier (from callback url): ')

    t = Tumblpy(key, secret, oauth_token, OAUTH_TOKEN_SECRET)

    authorized_tokens = t.get_authorized_tokens(oauth_verifier)

    final_oauth_token = authorized_tokens['oauth_token']
    final_oauth_token_secret = authorized_tokens['oauth_token_secret']

    print('OAuth Token: {}'.format(final_oauth_token))
    print('OAuth Token Secret: {}'.format(final_oauth_token_secret))
else:
    final_oauth_token = raw_input('OAuth Token: ')
    final_oauth_token_secret = raw_input('OAuth Token Secret: ')

t = Tumblpy(key, secret, final_oauth_token, final_oauth_token_secret)

blog_url = t.post('user/info')
blog_url = blog_url['user']['blogs'][0]['url']

print('Your blog url is: {}'.format(blog_url))

posts = t.posts(blog_url)

print('Here are some posts this blog has made:', posts)

# print t.post('post', blog_url=blog_url, params={'type':'text', 'title': 'Test', 'body': 'Lorem ipsum.'})


tumblpy/exceptions.py
class TumblpyError(Exception):
    """Generic error class, catch-all for most Tumblpy issues.
    from tumblpy import TumblpyError, TumblpyRateLimitError, TumblpyAuthError
    """
    def __init__(self, msg, error_code=None):
        self.error_code = error_code
        if error_code is not None:
            if error_code == 503:
                raise TumblpyRateLimitError(msg, error_code)
            elif error_code == 401:
                raise TumblpyAuthError(msg, error_code)

        super(TumblpyError, self).__init__(msg)

    @property
    def msg(self):
        return self.args[0]


class TumblpyRateLimitError(TumblpyError):
    """Raised when you've hit an API limit."""
    pass


class TumblpyAuthError(TumblpyError):
    """Raised when you try to access a protected resource and it fails due to
     some issue with your authentication."""
    pass


tumblpy.py
""" Tumblpy """

__author__ = 'Mike Helmick <mikehelmick@me.com>'
__version__ = '0.7.1'

import requests
from requests_oauthlib import OAuth1

try:
    import simplejson as json
except ImportError:
    try:
        import json
    except ImportError:
        try:
            from django.utils import simplejson as json
        except ImportError:
            raise ImportError('A json library is required to use this python library. Lol, yay for being verbose. ;)')

try:
    from urlparse import parse_qsl
except ImportError:
    from cgi import parse_qsl

import urllib


def _split_params_and_files(params_):
        params = {}
        files = {}
        for k, v in params_.items():
            if hasattr(v, 'read') and callable(v.read):
                files[k] = v
            elif isinstance(v, basestring):
                params[k] = v
            elif isinstance(v, bool):
                params[k] = 'true' if v else 'false'
            elif isinstance(v, int):
                params[k] = unicode(v)
            else:
                raise TumblpyError('Value for "%s" was not parsable.' % k)
        return params, files


class TumblpyError(Exception):
    """Generic error class, catch-all for most Tumblpy issues.
    from tumblpy import TumblpyError, TumblpyRateLimitError, TumblpyAuthError
    """
    def __init__(self, msg, error_code=None):
        self.msg = msg
        self.error_code = error_code
        if error_code is not None:
            if error_code == 503:
                raise TumblpyRateLimitError(msg, error_code)
            elif error_code == 401:
                raise TumblpyAuthError(msg, error_code)

    def __str__(self):
        return repr(self.msg)


class TumblpyRateLimitError(TumblpyError):
    """Raised when you've hit an API limit."""
    def __init__(self, msg, error_code=None):
        self.msg = msg
        self.error_code = error_code

    def __str__(self):
        return repr(self.msg)


class TumblpyAuthError(TumblpyError):
    """Raised when you try to access a protected resource and it fails due to
     some issue with your authentication."""
    def __init__(self, msg, error_code=None):
        self.msg = msg
        self.error_code = error_code

    def __str__(self):
        return repr(self.msg)


class TumblpyTimeout(TumblpyError):
    """Raised when you've set a timeout and this timeout has expired during the
    request."""
    pass


class Tumblpy(object):
    def __init__(self, app_key=None, app_secret=None, oauth_token=None, \
                oauth_token_secret=None, headers=None, callback_url=None, \
                pool_maxsize=None):

        # Define some API URLs real quick
        self.base_api_url = 'http://api.tumblr.com'
        self.api_version = 'v2'
        self.api_url = '%s/%s/' % (self.base_api_url, self.api_version)

        # Authentication URLs
        self.request_token_url = 'http://www.tumblr.com/oauth/request_token'
        self.access_token_url = 'https://www.tumblr.com/oauth/access_token'
        self.authorize_url = 'https://www.tumblr.com/oauth/authorize'
        self.authenticate_url = 'https://www.tumblr.com/oauth/authorize'

        self.callback_url = callback_url

        self.default_params = {'api_key': app_key}

        # If there's headers, set them, otherwise be an embarassing parent
        self.headers = headers or {'User-Agent': 'Tumblpy v' + __version__}

        if pool_maxsize:
            requests_config = {'pool_maxsize': pool_maxsize}
        else:
            requests_config = {}

        # Allow for unauthenticated requests
        self.client = requests.session()
        self.auth = None

        if app_key and app_secret:
            self.app_key = unicode(app_key) or app_key
            self.app_secret = unicode(app_secret) or app_key

        if oauth_token and oauth_token_secret:
            self.oauth_token = unicode(oauth_token)
            self.oauth_token_secret = unicode(oauth_token_secret)

        if app_key and app_secret and not oauth_token and not oauth_token_secret:
            self.auth = OAuth1(self.app_key, self.app_secret, signature_type='auth_header')

        if app_key and app_secret and oauth_token and oauth_token_secret:
            self.auth = OAuth1(self.app_key, self.app_secret,
                               self.oauth_token, self.oauth_token_secret,
                               signature_type='auth_header')

        if self.auth is not None:
            self.client.headers.update(self.headers)
            self.client.auth = self.auth

    def get_authentication_tokens(self):
        """ So, you want to get an authentication url?

            t = Tumblpy(YOUR_CONFIG)
            auth_props = t.get_authentication_tokens()
            auth_url = auth_props['auth_url']
            print auth_url
        """

        request_args = {}
        if self.callback_url:
            request_args['oauth_callback'] = self.callback_url

        response = self.client.get(self.request_token_url, params=request_args)

        if response.status_code != 200:
            raise TumblpyAuthError('Seems something couldn\'t be verified with your OAuth junk. Error: %s, Message: %s' % (response.status_code, response.content))

        request_tokens = dict(parse_qsl(response.content))
        if not request_tokens:
            raise TumblpyError('Unable to decode request tokens.')

        auth_url_params = {
            'oauth_token': request_tokens['oauth_token'],
        }
        if self.callback_url:
            auth_url_params['oauth_callback'] = self.callback_url

        request_tokens['auth_url'] = self.authenticate_url + '?' + urllib.urlencode(auth_url_params)

        return request_tokens

    def get_authorized_tokens(self, oauth_verifier):
        """Returns authorized tokens after they go through the auth_url phase.
        """
        response = self.client.get(self.access_token_url,
                                    params={'oauth_verifier': oauth_verifier})
        authorized_tokens = dict(parse_qsl(response.content))
        if not authorized_tokens:
            raise TumblpyError('Unable to decode authorized tokens.')

        return authorized_tokens

    def request(self, endpoint, method='GET', blog_url=None,
                extra_endpoints=None, params=None, timeout=None):
        params = params or {}
        method = method.lower()

        if not method in ('get', 'post'):
            raise TumblpyError('Method must be of GET or POST')

        url = self.api_url  # http://api.tumblr.com/v2/

        if blog_url is not None:
            # http://api.tumblr.com/v2/blog/blogname.tumblr.com/
            blog_url = blog_url.rstrip('/')
            if blog_url.startswith('http://'):
                blog_url = blog_url[7:]

            url = '%sblog/%s/' % (self.api_url, blog_url)

        url = '%s%s' % (url, endpoint)
        if extra_endpoints is not None:
            # In cases like:
            # http://api.tumblr.com/v2/blog/blogname.tumblr.com/posts/type/
            # 'type' is extra in the url & thought this was the best way
            # Docs: http://www.tumblr.com/docs/en/api/v2#posts

            url = '%s/%s' % (url, '/'.join(extra_endpoints))

        params, files = _split_params_and_files(params)
        params.update(self.default_params)

        func = getattr(self.client, method)
        try:
            if method == 'get':
                response = func(url, params=params, headers=self.headers,
                                timeout=timeout, allow_redirects=False)
            else:
                response = func(url,
                                data=params,
                                files=files,
                                headers=self.headers,
                                timeout=timeout,
                                allow_redirects=False)

        except requests.exceptions.Timeout:
            raise TumblpyTimeout('Request timed out.')

        except requests.exceptions.RequestException:
            raise TumblpyError('An unknown error occurred.')

        if response.status_code == 401:
            raise TumblpyAuthError('Error: %s, Message: %s' % (response.status_code, response.content))

        try:
            if endpoint == 'avatar':
                content = {
                    'response': {
                        'url': response.headers.get('location')
                    }
                }
            else:
                content = json.loads(response.content)
        except ValueError:
            raise TumblpyError('Unable to parse response, invalid JSON.')

        try:
            content = content.get('response', {})
        except AttributeError:
            raise TumblpyError('Unable to parse response, invalid content returned: %s' % content)

        if response.status_code < 200 or response.status_code > 301:
            error_message = ''
            if content and (content.get('errors') or content.get('error')):
                if 'errors' in content:
                    for error in content['errors']:
                        error_message = '%s ' % error
                elif 'error' in content:
                    error_message = content['error']

            error_message = error_message or \
                            'There was an error making your request.'
            raise TumblpyError(error_message, error_code=response.status_code)

        return content

    def get(self, endpoint, blog_url=None, extra_endpoints=None, params=None,
            timeout=None):
        return self.request(endpoint, blog_url=blog_url,
                            extra_endpoints=extra_endpoints, params=params,
                            timeout=timeout)

    def post(self, endpoint, blog_url=None, extra_endpoints=None, params=None,
             timeout=None):
        return self.request(endpoint, method='POST', blog_url=blog_url,
                            extra_endpoints=extra_endpoints, params=params,
                            timeout=timeout)

    def get_avatar_url(self, blog_url, size=64, timeout=None):
        size = [str(size)] or ['64']
        return self.get('avatar', blog_url=blog_url, extra_endpoints=size,
                        timeout=timeout)

    def __repr__(self):
        return u'<TumblrAPI: %s>' % self.app_key
